File: javascript/bracket-push/bracket-push.js
export const bracketPush = (s) => {
  const stack = [];
  const brackets = { '[': ']', '{': '}', '(': ')' };

  for (let i = 0; i < s.length; i++) {
    const c = s[i];
    if (Object.keys(brackets).includes(c)) {
      stack.push(brackets[c]);
    } else if (stack.pop() !== c) {
      return false;
    }
  }
  return stack.length === 0;
};

File: dart/bracket-push/lib/bracket_push.dart
class BracketPush {
  bool isPaired(String s) {
    List<String> stack = [];
    Map<String, String> brackets = { '{': '}', '[': ']', '(': ')' };
    for (var i = 0; i < s.length; i++) {
      var c = s[i];
      if (brackets.containsKey(c)) {
        stack.add(brackets[c]);
      } else if (brackets.containsValue(c)) {
        if (stack.removeLast() != c) {
          return false;
        }
      }
    }
    return stack.length == 0;
  }
}
